src/Symfony/Component/Uid/ArgumentResolver/UidValueResolver.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Uid\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Component\Uid\AbstractUid;
+
+final class UidValueResolver implements ArgumentValueResolverInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request, ArgumentMetadata $argument): bool
+    {
+        return !$argument->isVariadic() &&
+            \is_string($request->attributes->get($argument->getName())) &&
+            null !== $argument->getType() &&
+            is_subclass_of($argument->getType(), AbstractUid::class, true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        /** @var class-string<AbstractUid> $uidClass */
+        $uidClass = $argument->getType();
+
+        try {
+            return [$uidClass::fromString($request->attributes->get($argument->getName()))];
+        } catch (\InvalidArgumentException $e) {
+            throw new NotFoundHttpException(sprintf('The uid for the "%s" parameter is invalid.', $argument->getName()), $e);
```

Comment:
on contrary of the [enum resolver](https://github.com/symfony/symfony/pull/44831) for which we have a finite list of paths, I'm not convinced it should be a 404 here, but rather a bad request due to a malformed uid. WDYT?

Reply:
It's like having a `/users/{id}` route with a `\d+` requirement on the id and trying to access `/users/foo` for me. It's a 404 because the requirements for the route to "exist" are not met. So personally, I prefer a 404.

Reply:
A 404 is correct to me. The request is not bad per see, it's just that there is no content there. Aka 404 Not Found.